### PR TITLE
Criticality dev

### PR DESCRIPTION
**Pull request by Stern5497**
**What is new:**
- Code which can be used by multiple subclasses was moved to dataset_creator
- Citations in criticality_dataset_creator are weighted. The newer a citations is the more impotant it is. We use only citations from the last 21 years and normalize the weight by 21.
- Data directly from db is different to data from cache which caused some errors. This is solved by making sure some columns of dataframe from db has a certain type
- Canton, Chamber and Court are saved as a number in db, but we want a string representation, which was added.

**Important:**
- There were many changes to code of doc2doc_ir_dataset_creator -> needs to be tested.

### CODE REVIEW

File: scrc/dataset_creation/dataset_creator.py [hunk 1]
```diff
@@ -709,14 +827,15 @@ def split_random(self, df, split):
...
+        # TODO add secret_test
         train, val, test = dd.from_pandas(df, npartitions=1).random_split(list(split), random_state=self.seed)
 
         # get pandas dfs again
         train = train.compute(scheduler='processes')
         val = val.compute(scheduler='processes')
         test = test.compute(scheduler='processes')
 
-        return train, val, test
+        return train, val, test, []
```
Comment (JoelNiklaus) [inline scrc/dataset_creation/dataset_creator.py:838]: Is there a reason why the last value in the tuple is not a dataframe but a list?

File: scrc/dataset_creation/criticality_dataset_creator.py [hunk 2]
```diff
@@ -290,101 +261,65 @@ def process_citations(self, df):
...
+        if not self.count_all_cits:
+            # count citations for one bge only once in bger
+            df['ruling_citation'] = df['ruling_citation'].apply(lambda x: set(x))
+
+        # create counter out of list
+        df['ruling_citation'] = df['ruling_citation'].apply(lambda x: Counter(x))
+
+        # multiply weight to value in counter
+        def weight_citations(row):
+            weight = (row['year'] - 2001)
```
Comment (JoelNiklaus) [inline scrc/dataset_creation/criticality_dataset_creator.py:295]: Is it possible to replace this magic number with a constant or even compute it? (e.g., the lowest year in the df)

**Review by JoelNiklaus — COMMENTED** [review]
Looks great already, thank you!

I left you some suggestions to improve the code.

File: scrc/dataset_creation/dataset_creator.py [hunk 3]
```diff
@@ -155,6 +160,86 @@ def get_chunksize(self):
...
+                return result.cit_string()
+            return found_citation.cit_string()
+
+    def get_law_citation(self, citations_text):
+        """
+        handle single law citation
+        """
+        raise NotImplementedError("This method should be implemented in the subclass.")
+
     def create_dataset(self, sub_datasets=False, kaggle=False, huggingface=False, save_reports=False):
```
[the remark this answers is not in the document]
Comment (Stern5497) [reply scrc/dataset_creation/dataset_creator.py:243]: Which function do you mean? I guess you meant get_df()?

File: scrc/dataset_creation/dataset_creator.py [hunk 4]
```diff
@@ -309,15 +396,50 @@ def get_one_file_number(column_data):
...
+            if not math.isnan(float(x)):
+                return court_dict[int(x)]
+            else:
+                return np.nan
+
+        def get_chamber_value(x):
+            if not math.isnan(float(x)):
+                return chamber_dict[int(x)]
+            else:
+                return np.nan
```
[the remark this answers is not in the document]
Comment (Stern5497) [reply scrc/dataset_creation/dataset_creator.py]: Done

File: scrc/dataset_creation/dataset_creator.py [hunk 5]
```diff
@@ -457,15 +579,14 @@ def save_splits(self, splits: dict, labels: list, folder: Path,
     def create_splits(self, df, split, split_type, include_all=False):
         self.logger.info("Splitting data into train, val and test set")
         if split_type == "random":
-            train, val, test = self.split_random(df, split)
+            train, val, test, secret_test = self.split_random(df, split)
             secret_test = pd.DataFrame()
```
[the remark this answers is not in the document]
Comment (Stern5497) [reply scrc/dataset_creation/dataset_creator.py]: removed

File: scrc/dataset_creation/dataset_creator.py (hunk 1 above)
Comment (Stern5497) [reply scrc/dataset_creation/dataset_creator.py]: removed

File: scrc/dataset_creation/dataset_creator.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (JoelNiklaus) [reply scrc/dataset_creation/dataset_creator.py:243]: I actually meant the create_dataset()

[the remark this answers is not in the document]
Comment (JoelNiklaus) [reply scrc/dataset_creation/dataset_creator.py:243]: But it's ok, I can do it. Thanks!

**Pull request merged**